pkg/streamingpromql/operators/aggregation.go

```diff
@@ -136,14 +140,121 @@ func (a *Aggregation) SeriesMetadata(ctx context.Context) ([]types.SeriesMetadat
 	return seriesMetadata, nil
 }
 
-func (a *Aggregation) labelsForGroup(m labels.Labels, lb *labels.Builder) labels.Labels {
+// seriesToGroupLabelsStringFunc returns a function that computes a string representation of the output group labels for the given input series.
+//
+// The returned function returns a byte slice rather than a string to make it possible to avoid unnecessarily allocating a string.
+func (a *Aggregation) seriesToGroupLabelsStringFunc() func(labels.Labels) []byte {
+	// Why not just use the labels.Labels computed by seriesToGroupLabelsFunc() and call String() on it?
+	//
+	// Most of the time, we don't need the labels.Labels instance, as we expect there are far fewer output groups than input series,
+	// and we only need the labels.Labels instance once per output group.
+	// However, we always need to compute the string representation for each input series, so we can look up its corresponding
+	// output group. And we can do this without allocating a string by returning just the bytes that make up the string.
+	// There's not much point in using the hash of the group labels as we always need the string (or the labels.Labels) to ensure
+	// there are no hash collisions - so we might as well just go straight to the string representation.
+	//
+	// Furthermore, labels.Labels.String() doesn't allow us to reuse the buffer used when producing the string or to return a byte slice,
+	// whereas this method does.
+	// This saves us allocating a new buffer and string for every single input series, which has a noticeable performance impact.
+
+	if a.Without {
+		b := make([]byte, 0, 1024)
+		buf := bytes.NewBuffer(b)
+
+		return func(l labels.Labels) []byte {
+			buf.Reset()
+			nextGroupingIndex := 0
+			outputLabelCount := 0
```

Comment:
This doesn't need to be an int. Could use a bool here and set it to true after appending.

Reply:
Good catch, fixed in ce303e8